lib/matplotlib/colorbar.py

```diff
@@ -807,33 +805,50 @@ def _edges(self, X, Y):
             return [list(zip(Y[i], X[i])) for i in range(1, N - 1)]
 
     def _add_solids(self, X, Y, C):
-        """
-        Draw the colors using `~.axes.Axes.pcolormesh`;
-        optionally add separators.
-        """
-        if self.orientation == 'vertical':
-            args = (X, Y, C)
-        else:
-            args = (np.transpose(Y), np.transpose(X), np.transpose(C))
-        kw = dict(cmap=self.cmap,
-                  norm=self.norm,
-                  alpha=self.alpha,
-                  edgecolors='None')
-        _log.debug('Setting pcolormesh')
-        col = self.ax.pcolormesh(*args, **kw, shading='flat')
-        # self.add_observer(col) # We should observe, not be observed...
-
+        """Draw the colors; optionally add separators."""
+        # Cleanup previously set artists.
         if self.solids is not None:
             self.solids.remove()
-        self.solids = col
-
-        if self.drawedges:
-            self.dividers.set_segments(self._edges(X, Y))
+        for solid in self.solids_patches:
+            solid.remove()
+        # Add new artist(s), based on mappable type.  Use individual patches if
+        # hatching is needed, pcolormesh otherwise.
+        mappable = getattr(self, 'mappable', None)
+        if (isinstance(mappable, contour.ContourSet)
+                and any(hatch is not None for hatch in mappable.hatches)):
+            self._add_solids_patches(X, Y, C, mappable)
         else:
-            self.dividers.set_segments([])
+            self._add_solids_pcolormesh(X, Y, C)
+        self.dividers.set_segments(self._edges(X, Y) if self.drawedges else [])
+
+    def _add_solids_pcolormesh(self, X, Y, C):
+        _log.debug('Setting pcolormesh')
+        args = (X, Y, C) if self.orientation == 'vertical' else (Y.T, X.T, C.T)
+        self.solids = self.ax.pcolormesh(
+            *args, cmap=self.cmap, norm=self.norm, alpha=self.alpha,
+            edgecolors='none', shading='flat')
+        if not self.drawedges:
             if len(self._y) >= self.n_rasterize:
                 self.solids.set_rasterized(True)
 
+    def _add_solids_patches(self, X, Y, C, mappable):
+        hatches = mappable.hatches * len(C)  # Have enough hatches.
+        patches = []
+        for i in range(len(X) - 1):
+            xy = np.array([[X[i, 0], Y[i, 0]],
+                           [X[i, 1], Y[i, 0]],
+                           [X[i + 1, 1], Y[i + 1, 0]],
+                           [X[i + 1, 0], Y[i + 1, 1]]])
```

Comment:
is there code to factor out this building of a path for a patch? or would it be useful to have a make_rectangle_from_xy function?

Reply:
I wouldn't bother, there's too many different ways to represent x and y and providing conversions between all representations is not really worth it IMO (at least for this PR).

Reply:
I hear/not gonna block on it, I just have code that looks almost exactly like this somewhere 🙃